src/contracts/GPv2Settlement.sol

```diff
@@ -142,4 +150,57 @@ contract GPv2Settlement {
         );
         return IUniswapV2Pair(uint256(pairAddressBytes));
     }
+
+    /// @dev Verifies the specified clearing price is valid. Spefically, this
+    /// method verifies that the ±0.1% fee spread around the clearing price is
+    /// completely within the ±0.3% AMM fee spread around the Uniswap AMM price.
+    /// The AMM price is either defined by the token reserves (AKA spot price)
+    /// in the case the AMM is not traded with, or the actual swap amount (AKA
+    /// effective price) otherwise.
+    function verifyClearingPrice(
+        IUniswapV2Pair pair,
+        int112 d0,
+        int112 d1,
+        uint112 clearingPrice0,
+        uint112 clearingPrice1
+    ) internal view {
+        uint256 uniPrice0;
+        uint256 uniPrice1;
+        if (d0 | d1 == 0) {
+            // NOTE: In case the amounts swapped with the AMM is 0, we want to
+            // ensure that the clearing price with spread is within the AMM
+            // price spread.
+            (uint112 reserve0, uint112 reserve1, ) = pair.getReserves();
+            (uniPrice0, uniPrice1) = (uint256(reserve0), uint256(reserve1));
+        } else {
+            // NOTE: We want to check that exactly only one of `d0` or `d1` is
+            // strictly negative, and the other is strictly positive.
+            // SAFETY: `int112 * int112` cannot overflow a `int256`.
+            require(int256(d0) * int256(d1) < 0, "invalid Uniswap amounts");
+
+            // NOTE: Determine the effective AMM price **before fees**, this is
+            // required when verifying the clearing price as the direction of
+            // the trade is important here.
+            if (d0 < 0) {
+                (uniPrice0, uniPrice1) = (
+                    1000 * uint256(-d0),
+                    UNI_FEEFACTOR * uint256(d1)
```

Comment:
I am wondering, whether it would be better to assume that fees are already considered in d0 or d1 and the price is just `(uniPrice0, uniPrice1) = (-d0,d1)`

Kinda, uniswap contract is doing the same: https://github.com/Uniswap/uniswap-v2-core/blob/master/contracts/UniswapV2Pair.sol#L170. 

And it could save us some multiplications in many places. While it does not safe too much gas, it could increase readability.
But I am really not sure about what is best.

Reply:
The reason I "undo" the fees here is so the calculation for whether the clearing price + spread falls withing the AMM price + spread was the same whether the Uniswap price is the spot price or the effective price.

If we don't want a price range when we trade with Uniswap, then this becomes simpler and using your suggestion makes a lot of sense.

I'll change this to :point_up: for now, and if we decide we want to use a price range when we go to Uniswap, we can reconsider then.

Reply:
Oh right, you are right. Let's keep it as it was...